stashcp/__init__.py

```diff
@@ -559,100 +563,184 @@ def get_ips(name):
+        api_text = "stashservers.dat"
+        if cache_list_name != None:
+            api_text += '?list=' + cache_list_name
+
+    responselines = []
     i = 0
-    while order_str == '' and i < len(geo_ip_sites):
+    while len(responselines) == 0 and i < len(geo_ip_sites):
         cur_site = geo_ip_sites[i]
         headers['Host'] = cur_site
+        logging.debug("Trying server site of %s", cur_site)
         for ip in get_ips(cur_site):
-            logging.debug("Trying geoip site of: %s [%s]", cur_site, ip)
-            final_url = "http://%s/%s/%s" % (ip, append_text, caches_string)
-            logging.debug("Querying for closest cache: %s", final_url)
+            final_url = "http://%s/%s" % (ip, api_text)
+            logging.debug("Querying %s", final_url)
             try:
                 # Make the request
                 req = urllib2.Request(final_url, headers=headers)
                 response = urllib2.urlopen(req, timeout=10)
                 if response.getcode() == 200:
                     logging.debug("Got OK code 200 from %s", cur_site)
-                    order_str = response.read()
+                    responselines = response.read().split('\n')
                     response.close()
                     break
                 response.close()
             except urllib2.URLError, e:
                 logging.debug("URL error: %s", str(e))
             except Exception, e:
                 logging.debug("Error: %s", str(e))
-            i+=1
+        i+=1
+
+    order_str = ''
+    if len(responselines) > 0:
+        order_str = responselines[0]
         
     if order_str == '':
+        if len(caches_list) == 0:
+            logging.error("unable to get list of caches")
+            return None
         # Unable to find a geo_ip server to use, return random choice from caches!
-        minsite = random.choice(caches_list)['name']
-        random.shuffle(caches_list)
-        nearest_cache_list = [cache['name'] for cache in caches_list]
+        nearest_cache_list = caches_list
+        random.shuffle(nearest_cache_list)
+        minsite = nearest_cache_list[0]
         logging.warning("Unable to use Geoip to find closest cache!  Returning random cache %s", minsite)
-        logging.debug("Ordered list of nearest caches: %s", str(nearest_cache_list))
+        logging.debug("Randomized list of nearest caches: %s", str(nearest_cache_list))
         return minsite
     else:
         # The order string should be something like:
         # 3,1,2
         ordered_list = order_str.strip().split(",")
         logging.debug("Got order %s", str(ordered_list))
-        minsite = caches_list[int(ordered_list[0])-1]['name']
+
+        if len(caches_list) == 0:
+            # Used the stashservers.dat api
+
+            if len(responselines) < 8:
+                logging.error("stashservers response too short, less than 8 lines")
+                return None
+            hashname = responselines[4][-5:]
```

Comment:
To make this more illustrative, could you put an example of this file's contents in a comment near here.  So we can see what lines are what?